lib/AST/GenericSignatureBuilder.cpp

```diff
@@ -3910,6 +4592,20 @@ GenericSignatureBuilder::addSameTypeRequirementBetweenArchetypes(
                                    equivClass->sameTypeConstraints.end(),
                                    equivClass2->sameTypeConstraints.begin(),
                                    equivClass2->sameTypeConstraints.end());
+
+    // Combine the rewrite rules.
+    if (auto rewriteRoot2 = Impl->getOrCreateRewriteTreeRoot(equivClass2)) {
+      if (auto rewriteRoot1 = Impl->getOrCreateRewriteTreeRoot(equivClass)) {
+        // Merge the second rewrite tree into the first.
+        rewriteRoot2->mergeInto(rewriteRoot1);
+        Impl->RewriteTreeRoots.erase(equivClass2);
+        delete rewriteRoot2;
```

Comment:
I'm curious: since there's at least two of these manual clean-ups, `unique_ptr` or similar mustn't work with `DenseMap` (and how it's being used); why?

Reply:
I was thinking of doing something clever with re-using root nodes, but I don't think it's worthwhile.  I'll switch it over to using `std::unique_ptr` for the roots.